File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.java
import java.util.*;
class Solution {
    public int solution(int cacheSize, String[] cities) {
        int answer = 0;
        
        List<String> arr = new ArrayList<>();
        
        if (cacheSize == 0) return cities.length * 5;
        
        for (String city : cities) {
            city = city.toLowerCase();
            
            if (arr.contains(city)) {
                arr.remove(city);
                arr.add(city);
                answer += 1;
            } else {
                if (arr.size() == cacheSize) {
                    arr.remove(0);
                    arr.add(city);
                    answer += 5;
                } else {
                    arr.add(city);
                    answer += 5;
                }
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.cpp
#include <bits/stdc++.h>

using namespace std;

deque<string> dq;

int solution(int cacheSize, vector<string> cities) {
    int answer = 0;

    for (string city : cities) {
        if (cacheSize == 0) {
            answer += 5;
            continue;
        }
        
        for (int i=0;i<city.size();i++) {
            city[i] = tolower(city[i]);
        }
        int ok = 0;
        for (int i=0;i<dq.size();i++) {
            if (dq[i] == city) {
                // 캐시히트
                ok = 1;
                dq.erase(dq.begin() + i);
                dq.push_front(city);
                answer += 1;
                break;
            }
        }
        if (!ok) {
            if (dq.size() < cacheSize) {
                dq.push_front(city);
                answer += 5;
            } else {
                dq.pop_back();
                dq.push_front(city);
                answer += 5;
            }
        }
    }

    return answer;
}